File: app/app/src/test/java/dev/steamvault/app/ui/library/logic/LibraryFiltersTest.kt
package dev.steamvault.app.ui.library.logic

import dev.steamvault.app.net.model.GameSummary
import dev.steamvault.app.net.model.JobSummary
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

private fun game(appid: Int, name: String, status: String = "idle", sizeBytes: Long? = null): GameSummary =
    GameSummary(
        appid = appid,
        name = name,
        status = status,
        last_prefill_at = null,
        last_manifest_check = null,
        depot_count = 0,
        size_bytes = sizeBytes,
        needs_force = false,
    )

private fun job(appid: Int, status: String = "running"): JobSummary = JobSummary(
    id = appid,
    appid = appid,
    type = "prefill",
    status = status,
    created_at = "2026-08-01T00:00:00Z",
)

class LibraryFiltersTest {

    private val notCached = game(1, "Nebula Drift")
    private val cached = game(2, "Ironwood Hollow", status = "done", sizeBytes = 5_000_000_000)
    private val failed = game(3, "Tundra Protocol", status = "error")
    private val downloading = game(4, "Meridian Rally")

    private val games = listOf(notCached, cached, failed, downloading)
    private val liveJobsByAppid = indexLiveJobsByAppid(listOf(job(4, "running")))

    // ---- normalizeQuery / matchesQuery ----------------------------------

    @Test
    fun `normalizeQuery trims and lowercases`() {
        assertEquals("nebula", normalizeQuery("  Nebula  "))
        assertEquals("", normalizeQuery(null))
    }

    @Test
    fun `matchesQuery is case-insensitive substring (query pre-normalized by the caller, mirrors normalizeQuery contract)`() {
        assertTrue(matchesQuery(notCached, "drift"))
        assertTrue(matchesQuery(notCached, normalizeQuery("NEBULA")))
        assertFalse(matchesQuery(notCached, "hollow"))
    }

    @Test
    fun `matchesQuery always true for an empty query`() {
        assertTrue(matchesQuery(notCached, ""))
    }

    // ---- the recorded chip set: All Cached Not-cached Downloading Failed --

    @Test
    fun `chip set is exactly All Cached Not-cached Downloading Failed, in order`() {
        assertEquals(
            listOf("all", "cached", "none", "downloading", "failed"),
            FILTER_DEFS.map { it.key },
        )
    }

    @Test
    fun `filterByKey falls back to all for an unknown key`() {
        assertEquals("all", filterByKey("update_ready").key)
    }

    @Test
    fun `chip set labels match the web port's FILTER_DEFS labels verbatim`() {
        // Literal-vs-literal pin (docs/LEARNINGS.md "Android (Phase 4b)":
        // cross-frontend contracts need LITERAL expected sets, never
        // derived from the enum/list under test) -- hand-transcribed from
        // web/js/lib/library-filters.js's own FILTER_DEFS, not read back
        // from this file. This is what qualifies FILTER_DEFS's labels for
        // the literal-Kotlin-string exception this file's kdoc documents,
        // instead of routing them through strings.xml.
        assertEquals(
            listOf("All", "Cached", "Not cached", "Downloading", "Failed"),
            FILTER_DEFS.map { it.label },
        )
    }

    // ---- visibleGames: search AND chip -----------------------------------

    @Test
    fun `all chip with no query returns everything`() {
        val visible = visibleGames(games, query = "", filterKey = "all", liveJobsByAppid = liveJobsByAppid)
        assertEquals(games.map { it.appid }, visible.map { it.appid })
    }

    @Test
    fun `cached chip returns only cached-with-bytes games`() {
        val visible = visibleGames(games, query = "", filterKey = "cached", liveJobsByAppid = liveJobsByAppid)
        assertEquals(listOf(cached.appid), visible.map { it.appid })
    }

    @Test
    fun `none chip returns only not-cached games`() {
        val visible = visibleGames(games, query = "", filterKey = "none", liveJobsByAppid = liveJobsByAppid)
        assertEquals(listOf(notCached.appid), visible.map { it.appid })
    }

    @Test
    fun `downloading chip returns only games with a live job`() {
        val visible = visibleGames(games, query = "", filterKey = "downloading", liveJobsByAppid = liveJobsByAppid)
        assertEquals(listOf(downloading.appid), visible.map { it.appid })
    }

    @Test
    fun `failed chip returns only error-status games`() {
        val visible = visibleGames(games, query = "", filterKey = "failed", liveJobsByAppid = liveJobsByAppid)
        assertEquals(listOf(failed.appid), visible.map { it.appid })
    }

    @Test
    fun `search and chip are ANDed`() {
        val visible = visibleGames(
            games,
            query = "tundra",
            filterKey = "failed",
            liveJobsByAppid = liveJobsByAppid,
        )
        assertEquals(listOf(failed.appid), visible.map { it.appid })

        val emptyIntersection = visibleGames(
            games,
            query = "tundra",
            filterKey = "cached",
            liveJobsByAppid = liveJobsByAppid,
        )
        assertTrue(emptyIntersection.isEmpty())
    }

    // ---- chipCounts: recompute against the current query -----------------

    @Test
    fun `chipCounts against an empty query count the whole library`() {
        val counts = chipCounts(games, query = "", liveJobsByAppid = liveJobsByAppid).associateBy { it.key }
        assertEquals(4, counts.getValue("all").count)
        assertEquals(1, counts.getValue("cached").count)
        assertEquals(1, counts.getValue("none").count)
        assertEquals(1, counts.getValue("downloading").count)
        assertEquals(1, counts.getValue("failed").count)
    }

    @Test
    fun `chipCounts narrow to only what the current query would show`() {
        val counts = chipCounts(games, query = "meridian", liveJobsByAppid = liveJobsByAppid).associateBy { it.key }
        assertEquals(1, counts.getValue("all").count)
        assertEquals(0, counts.getValue("cached").count)
        assertEquals(1, counts.getValue("downloading").count)
        assertEquals(0, counts.getValue("failed").count)
    }
}

File: web/tests/library-filters.test.js
/**
 * Headless tests for web/js/lib/library-filters.js (WP 4a.3).
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  matchesQuery,
  normalizeQuery,
  visibleGames,
  chipCounts,
  FILTER_DEFS,
} from "../js/lib/library-filters.js";
import { indexLiveJobsByAppid } from "../js/lib/game-status.js";

const game = (over) => ({
  appid: 1,
  name: "Aurora Cascade",
  status: "idle",
  last_prefill_at: null,
  size_bytes: null,
  ...over,
});

const GAMES = [
  game({ appid: 1, name: "Aurora Cascade", status: "done", size_bytes: 5_000_000_000 }), // cached
  game({ appid: 2, name: "Copper Horizon", status: "idle", size_bytes: null }), // none
  game({ appid: 3, name: "Driftwood Signal", status: "error", size_bytes: null }), // failed
  game({ appid: 4, name: "Ashfall Requiem", status: "idle", size_bytes: null }), // none
];
const JOBS = [{ id: 900, appid: 4, type: "prefill", status: "running" }];
const LIVE = indexLiveJobsByAppid(JOBS);

test("matchesQuery: case-insensitive substring on the title", () => {
  // matchesQuery's contract (see its docstring) is that `query` arrives
  // ALREADY normalized (normalizeQuery, tested below) — the case
  // insensitivity this test pins is the GAME NAME's casing, not the
  // query's; query casing is the caller's job.
  assert.equal(matchesQuery(game({ name: "Aurora Cascade" }), "cascade"), true);
  assert.equal(matchesQuery(game({ name: "AURORA CASCADE" }), "cascade"), true);
  assert.equal(matchesQuery(game({ name: "Aurora Cascade" }), "zzz"), false);
});
test("matchesQuery: an empty query matches everything", () => {
  assert.equal(matchesQuery(game({ name: "anything" }), ""), true);
});

test("normalizeQuery trims and lowercases", () => {
  assert.equal(normalizeQuery("  Cascade  "), "cascade");
  assert.equal(normalizeQuery(null), "");
  assert.equal(normalizeQuery(undefined), "");
});

test("FILTER_DEFS has no 'stale'/'update ready' chip (no oracle data on GET /v1/games yet)", () => {
  assert.equal(
    FILTER_DEFS.some((f) => f.key === "stale" || f.key === "upd"),
    false,
  );
});

test("visibleGames: 'all' with no query returns everything", () => {
  const list = visibleGames(GAMES, { query: "", filterKey: "all", liveJobsByAppid: LIVE });
  assert.equal(list.length, 4);
});

test("visibleGames: search and chip are ANDed", () => {
  // "a" matches Aurora Cascade, Ashfall Requiem, Driftwood Signal(no) —
  // 'a' appears in "Ashfall" and "Aurora" and "Driftwood Signal" (has an
  // 'a' too)... pick an unambiguous substring instead.
  const list = visibleGames(GAMES, { query: "ash", filterKey: "none", liveJobsByAppid: LIVE });
  // Only Ashfall Requiem matches BOTH "contains 'ash'" AND "is Not cached"
  // (appid 4 is actually mid-download -> its filter kind is "downloading",
  // not "none" -- see the next test for that nuance).
  assert.deepEqual(list.map((g) => g.appid), []);
});

test("visibleGames: a live job moves a game OUT of the 'none' filter and INTO 'downloading'", () => {
  const asNone = visibleGames(GAMES, { query: "", filterKey: "none", liveJobsByAppid: LIVE });
  assert.ok(!asNone.some((g) => g.appid === 4), "appid 4 has a running job, so it is not 'none'");

  const asDownloading = visibleGames(GAMES, {
    query: "",
    filterKey: "downloading",
    liveJobsByAppid: LIVE,
  });
  assert.deepEqual(asDownloading.map((g) => g.appid), [4]);
});

test("visibleGames: 'cached' and 'failed' chips", () => {
  const cached = visibleGames(GAMES, { query: "", filterKey: "cached", liveJobsByAppid: LIVE });
  assert.deepEqual(cached.map((g) => g.appid), [1]);
  const failed = visibleGames(GAMES, { query: "", filterKey: "failed", liveJobsByAppid: LIVE });
  assert.deepEqual(failed.map((g) => g.appid), [3]);
});

test("chipCounts recompute against the CURRENT query — always what the grid can show", () => {
  const counts = chipCounts(GAMES, { query: "driftwood", liveJobsByAppid: LIVE });
  const byKey = Object.fromEntries(counts.map((c) => [c.key, c.count]));
  assert.equal(byKey.all, 1);
  assert.equal(byKey.failed, 1);
  assert.equal(byKey.cached, 0);
  assert.equal(byKey.none, 0);
});

test("chipCounts with no query sums to the full library for 'all'", () => {
  const counts = chipCounts(GAMES, { query: "", liveJobsByAppid: LIVE });
  const all = counts.find((c) => c.key === "all");
  assert.equal(all.count, GAMES.length);
});
